contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -4205,4 +4205,98 @@ LANGUAGE plpgsql STABLE;
 CREATE OR REPLACE FUNCTION sys.bbf_pivot()
 RETURNS setof record
 AS 'babelfishpg_tsql', 'bbf_pivot'
-LANGUAGE C STABLE;
\ No newline at end of file
+LANGUAGE C STABLE;
+
+-- wrapper functions for reverse
+CREATE OR REPLACE FUNCTION sys.reverse(string ANYELEMENT)
+RETURNS sys.VARCHAR
+AS
+$BODY$
+DECLARE
+    string_arg_datatype text;
+    string_basetype oid;
+BEGIN
+    string_arg_datatype := sys.translate_pg_type_to_tsql(pg_typeof(string)::oid);
+    IF string_arg_datatype IS NULL THEN
+        -- for User Defined Datatype, use immediate base type to check for argument datatype validation
+        string_basetype := (SELECT typbasetype FROM pg_type WHERE oid = pg_typeof(string)::oid);
+        string_arg_datatype := sys.translate_pg_type_to_tsql(string_basetype);
+    END IF;
+
+    -- restricting arguments with invalid datatypes for reverse function
+    IF string_arg_datatype IN ('image', 'sql_variant', 'xml', 'geometry', 'geography') THEN
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 1 of reverse function.', string_arg_datatype;
+    END IF;
+
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.reverse(string::sys.varchar);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;
+
+CREATE OR REPLACE FUNCTION sys.reverse(string sys.BPCHAR)
+RETURNS sys.VARCHAR
+AS
+$BODY$
+BEGIN
+    RETURN PG_CATALOG.reverse(string);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE OR REPLACE FUNCTION sys.reverse(string sys.VARCHAR)
+RETURNS sys.VARCHAR
```

Comment:
For this and the below function signatures, input type varchar, nvarchar, nchar..the function body is same. How are we making sure that for the specified input datatype, expected function is getting picked up? Is there any changes we have to make sure that happens? I think, for all these input datatypes, function with TEXT as input will be invoked. Can you please clarify it and if possible add tests to confirm the same?

Reply:
1. Underlying Postgres logic to choose the best candidate is as follows:
     1.1. first find the number of candidates whose definition(namespace, function name, function argument types) are matching exactly.
     1.2. if only one candidate found Postgres will use that, else Postgres uses some best match heuristics to find the best candidate.
This logic will make sure that expected function is getting picked up.

2. For function call with input argument types nchar, nvarchar, text and ntext its corresponding definition will be picked up, and for all other datatypes anyelement definition will be picked up.
3. Already added enough tests to confirm the same in reverse-vu-prepare, reverse-vu-verify and reverse-vu-cleanup testfiles.